[23기 이정일] [MODIFY]PR review 수정사항 반영 / [ADD] ShopListMain 컴포넌트에 '더보기' 기능 추가

201steve opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 2021.08.03 공통 Nav 컴포넌트 레이아웃 생성 완료
- 2021.08.03 ShopList페이지 Header 레이아웃 완료
- 2021.08.05 '더보기'기능 추가 
- 2021.08.09 '더보기'기능 수정, '찜' 기능추가 및 오류 수정 
- 2021.08.09 conflict 코드 수정
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 2021.08.03 공통 Nav 컴포넌트 레이아웃 완료 했는데, fontawesome kit script는 제외하고 commit 함(중복 사용으로 conflict 방지)
- 2021.08.03 ShopList 는 사용자가 원하는 메뉴나 키워드로 검색 된 가게 리스트를 보는 페이지로, Shopdetail 로 가기전 중간 페이지입니다.  
레이아웃 완성한 header에는 sorting된 shop list의 클릭 수, 오늘날짜, shop list의 타이틀과 서브타이틀을 화면에 출력합니다.
-2021.08.05 ShopListMain 컴포넌트에 Review의 크기를 전부 볼 수 있게 혹은 일부만 보이게 하는 기능을 추가 했습니다.
-2021.08.05 scss 리뷰대로 뒤죽박죽이던 속성 순서를 비슷한 순서로 정리했습니다.
-2021.08.09 모든 컴포넌트에 동시에 적용되던 더보기,찜 기능 버튼들을 각 컴포넌트 단위로만 동작하도록 코드 수정했습니다.
-2021.08.09 footer scss를 bottom-up방식으로 수정했습니다.
-2021.08.09 conflict 된 index.html 파일, 코드 수정해서 커밋 했습니다.

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 함수 로직이 너무 긴 것 같습니다. 좀 더 효율적인 방법이 없을지 궁금합니다.(예시) 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

201steve committed:
[ADD] 1.공통 Nav 레이아웃 완료, 2.ShopList Header 레이아웃 완료

Tonyk0901 commented on src/components/Nav/Nav.js:
```diff
@@ -1,10 +1,50 @@
 class Nav extends React.Component {
   render() {
     return (
       <div className="Nav">
```
- 대문자로 첫 글자를 시작하는 건, 컴포넌트 이름을 네이밍 할때만 입니다~ 클래스네임을 짓는 건 보통의 카멜케이스로 소문자로 시작해주세요!
```suggestion
      <div className="nav">
```
- 근데 사실, 이 div 태그는 불필요한 것 같습니다. 추가적인 스타일을 부여하기 위해서 사용하고 계신 것도 아니고, 가장 바깥에서 한가지 태그로 감싸기 위해서 사용하는 것도 아닌 것 같아서요! (이미 nav 태그로 가장 바깥에서 감싸주고 계십니다!)

201steve replied:
네!! 소문자로 수정했습니다!

Tonyk0901 commented on src/components/Nav/Nav.scss:
```diff
@@ -1,3 +1,103 @@
+  z-index: 1;
+  border: 0;
+  box-shadow: rgba(0, 0, 0, 0.1) 0px 4px 11px 0px;
+
```
- 가장 바깥에서 .nav 로 네스팅 해주세요!

201steve replied:
.nav 로 네스팅 바꿧습니다!

Tonyk0901 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,26 @@
 class ShopList extends React.Component {
   render() {
     return (
       <div className="ShopList">
```
- 여기도 네이밍 소문자로 시작해주세요!

201steve replied:
넵!!

Tonyk0901 commented on src/pages/ShopList/ShopList.scss:
```diff
@@ -1,3 +1,18 @@
-* {
-  width: 100%;
+.shopContainer {
```
- .shopList 로 가장 바깥에서 네스팅 해주세요!

201steve replied:
.shopList로 수정했습니다!

Tonyk0901 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,26 @@
+        <Nav />
+        <ShopListHeader />
+        <main>
+          <ul className="shopContainer">
+            <li className="shopList">
```
- 여러가지의 리스트가 있는 것이 아니기 때문에 ul + li 를 사용하는 건 자연스러워 보이지 않습니다! 
- 이후에 추가적인 내용물들이 들어올 예정인가요?

201steve replied:
내용들은 수정했습니다!

Tonyk0901 reviewed (commented):
고생하셨습니다 정일님! 크게 리뷰가 필요한 부분은 없고, 컨벤션적인 것들 몇개 리뷰 드렸는데 확인해주세요!

- css 속성 순서가 좀 뒤죽박죽인 것 같습니다! 리팩토링 체크리스트 확인하셔서 css 속성 순서 일관되게 수정해주세요!
- 그리고 css 에서 새로운 셀렉터를 시작하실 땐, 한칸 띄워주시기 바랍니다~

201steve committed:
23기 이정일 [ADD]공통 footer 레이아웃 완료 / shopList header,main 레이아웃 완료 /[fix]PR review 수정사항 반영

Tonyk0901 commented on src/components/Footer/Footer.js:
```diff
@@ -1,10 +1,78 @@
+          <div className="footerMiddle">
+            <div className="goodPlace">
+              <span>
+                근처맛집: 노스트레스버거 | 바스버거 | 하동관 | 은희네해장국 |
```
-🙅‍♂️ 😣  🍔  👍

Tonyk0901 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,20 @@
 
 class ShopList extends React.Component {
   render() {
     return (
-      <div className="ShopList">
-        <div>ShopList 입니다</div>
+      <div className="shopList">
+        <Nav />
+        <ShopListHeader />
+        <ShopListMain />
+        <ShopListMain />
+        <ShopListMain />
+        <ShopListMain />
+        <Footer />
       </div>
     );
   }
```
- 잘 해주셨습니다-! 만,
- 현재는 UI 를 그 안에 들어가는 데이터를 직접 작성하셔서 그려주고 계십니다. 
- 하지만 ShopListMain 안에 들어가는 정보들은 사실 백엔드에서 받아오는 부분이기 때문에, 이것들을 목업 데이터로 만들어서 그것을 통해 UI 를 구성하는 방식으로 수정해 주시면 더 좋을 것 같습니다. 그래야 이후에 백엔드 와 맞춰보실 때 더 수월하게 작업을 하실수 있습니다.
```jsx
class ShopList extends React.Componet {
  state = {
    data: [],
  }

  componentDidMount() {
    fetch('api')
      .then(res => res.json())
      .then(res => this.setState({
        data: res,
      }))
  }
  render() {
    return (
      <div className='shopList'>
        <Nav />
        <ShopListHeader />
        {this.state.data.map(itemList => (
          <ShopListMain itemList={itemList} />
        ))}
    )
  }
}
```
- 이렇게 되면, ShopListMain 에서는 UI에 그리는 데이터를 props 로 받도록 수정을 할 필요가 있겠네요!

201steve replied:
넹!! API로 데이터 받아서 하기전까지 레이아웃 확인용으로 만들어봤슴다! 
실제로 구현할 때는 map 써서 만들겠슴다!!

Tonyk0901 reviewed (commented):
정일님-! 백엔드에서 데이터를 받아와서 구성해야 하는 UI 의 경우는, 하드 코딩 하지마시고 목업 데이터를 구성하셔서 만들어주세요!

201steve committed:
[ADD] ShopListMain 컴포넌트에 '더보기' 기능 추가

lang92 commented on src/components/Footer/Footer.js:
```diff
@@ -1,10 +1,78 @@
+              <img
+                className="footerLogo"
+                src="images/tangoGreyLogo.svg"
+                alt=""
```
- img에 alt 값 제대로 부여해주시기 바랍니다! img 태그에 들어가는 alt 속성은 아래와 같은 이유로 매우 중요합니다(그렇기 때문에 src속성보다 먼저 작성하는게 좋습니다).

> 이미지가 로딩되지 않을 경우 alt 속성에 적힌 값이 보여짐.
시각장애인의 경우 alt 속성을 통해 어떤 이미지가 보여지는지 알 수 있음.
SEO 검색엔진 최적화에서 어떤 img인지 인식하는 코드는 alt 속성임.

201steve replied:
alt값 수정했습니다!

saengmotmi commented on src/components/Footer/Footer.scss:
```diff
@@ -1,3 +1,104 @@
-* {
-  width: 100%;
+.footer {
+  height: 650px;
+  background-color: #3e3e3e;
+
+  .footerContainer {
+    padding-top: 65px;
+    margin: 0 20px;
+
+    .footerTop {
+      height: 310px;
```
- 지금 레이아웃 잡는 방식을 보시면 매 요소마다 고정 가로 / 세로 높이를 전부 주고 계시네요. 하지만 생각해보면 자식 요소가 300px 이라고 가정했을 때 그 요소를 감싸고 있는 부모 요소는 당연히 최소 300px 일 거라고 생각해볼 수 있습니다.
- 또 하나 가정해보겠습니다. 예컨대 부모 요소(카드 컴포넌트)의 고정 높이가 300px 인데, 자식 요소의 내용이 데이터에 따라 많아져서 300px 높이를 초과하게 되었습니다. 만약 부모 요소에 고정 높이를 주지 않았다면 데이터에 맞도록 부모 요소의 크기가 조절되었을 텐데, 굳이 부모 요소에 고정 높이를 줘버린 탓에 레이아웃이 깨지게 되겠네요 🤔 
- 따라서 가급적 자식 요소 + padding, margin으로 레이아웃을 잡아 부모 요소는 자식 요소를 통해 크기가 결정될 수 있도록 해야겠습니다. 현재 레이아웃 잡으신 방식 일정 참고하셔서 제가 말씀드린 방식 대로 가급적 바꾸어주시되, 무리가 되겠다 싶으면 새로 작성하는 코드부터는 반드시 반영하여 작성해주세요.
- 만약 제가 말씀드린 리뷰 내용이 이해가 안된다면 다시 질문 주세요. 중요한 부분입니다.

201steve replied:
리뷰 감사합니다 :-) 말씀하신부분 어떤내용인지 확인했습니다 자식태그 크기만큼 부모태그 크기가 정해지도록 부모태그에  margin, padding이용해서 레이아웃 구성하는 방법으로 리팩토링 해보겠습니다. 감사합니다!

saengmotmi commented on src/components/Nav/Nav.scss:
```diff
@@ -1,3 +1,112 @@
+      }
+    }
+
+    .firstChild {
+      position: relative;
+      img {
+        width: 15px;
+        position: fixed;
+        top: 10px;
+        right: 35px;
+        position: absolute;
+      }
```
- first, second 등의 순서를 기반으로 셀렉터 이름을 지어주는건 이상합니다. 경우에 따라 순서는 충분히 변경될 수도 있을텐데, 그 때마다 모든 요소들의 이름을 다시 지어줄 수는 없는 노릇입니다
- position Fixed는 모든 규칙들을 무시하고 body 태그에 달라 붙는 효과가 있습니다. img 태그가 부모에 가서 붙을 텐데 .firstChild의 position, img 태그의 position 속성은 의미가 없어보입니다. 의도하신 부분이 맞는지 확인해주세요. 일단 img 태그에 position이 중복입니다.

201steve replied:
1. firstchild는 수정했습니다. 이건 처음에 네이밍할때 너무 간단히 생각했던부분이라 조금 민망하기까지 하네요. 네이밍 신경써서 하겠습니다 :-) 
2. position:abolute로 변경하고 position:fixed는 지웠어야 했는데 제가 고 부분 살짝 놓쳤습니다. 말씀해주셔서 감사합니다 :-)
코드에는 수정해서 반영했습니다!

saengmotmi commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,44 @@
+      <div className="shopList">
+        <Nav />
+        <ShopListHeader />
+        <ShopListMain
+          moreReview={this.state.moreReview}
+          buttonHandle={this.buttonHandler}
+          close={this.state.closeReview}
+        />
```
Props로 넘겨주고 있는 부분들이 ShopListMain에서만 사용되고 있지만 ShopList에서 만들고 내려주는 것이 아니라 아예 ShopListMain로 자리를 옮기는게 맞지 않을까요?

201steve replied:
아!! 모든 자식태그에 요소들이 부모컴포넌트에서 props로 받아서 써야 하는줄로만 알고있었습니다.
이건 제가 잘못 이해하고 있던부분이네요. 감사합니다 코드 수정해서 반영하겠습니다!

saengmotmi commented on src/pages/ShopList/ShopListMain.js:
```diff
@@ -0,0 +1,58 @@
+                </p>
+                <button
+                  className="moreReviewButton"
+                  onClick={this.props.buttonHandle}
```
구조 분해 할당 적용해주세요!

201steve replied:
네! 구조분해할당 적용하겠습니다!

saengmotmi reviewed (commented):
레이아웃 잡는 방식에 조금 수정할 부분이 있습니다. 리뷰 내용 확인해주시고 이해 안되는 부분은 질문 주세요.

lang92 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,44 @@
+    };
+  }
+
+  buttonHandler = () => {
```
함수명은 함수의 의도와 기능을 명확히 할 수 있는 동사형이어야 합니다.  ex) handleMoreInfoBtn

201steve replied:
동사형으로 수정했습니다!

lang92 commented on src/pages/ShopList/ShopListHeader.scss:
```diff
@@ -0,0 +1,42 @@
+
+    .innerWrapper {
+      height: 100%;
+      padding: 48px 30px 10px 30px;
```
```suggestion
      padding: 48px 30px 10px;
```
위와 같이 쓸 수 있습니다.

https://developer.mozilla.org/ko/docs/Web/CSS/padding

201steve replied:
중복되는 값 수정했습니다!

201steve committed:
[MODIFY] 리뷰 수정사항 반영
-footer 레이아웃 bottom-up방식으로 scss 수정
-선택자 변경

[FIX] 더보기,Like 토글 수정
-모든 컴포넌트에 동시 적용->컴포넌트마다 개별 작동하도록 로직 수정

[REFACTOR] 구조분해할당 적용
-state,props 구조분해할당 해서 중복 코드 제거

lang92 commented on src/components/Footer/Footer.scss:
```diff
@@ -1,3 +1,108 @@
+        .fa-facebook-f {
+          border: 1px solid #9b9b9b;
+          border-radius: 50%;
+          padding: 15px 17px 15px 17px;
```
```suggestion
          padding: 15px 17px;
```
위와 같이 쓸 수 있습니다. 리뷰 드리지 않은 다른 부분도 전체적으로 확인하고 수정해주세요.

lang92 commented on src/Routes.js:
```diff
@@ -10,7 +10,7 @@ class Routes extends React.Component {
+          <Route exact path="/" component={Main} />
           <Route exact path="/FilterList" component={FilterList} />
           <Route exact path="/ShopList" component={ShopList} />
           <Route exact path="/ShopDetail" component={ShopDetail} />
```
일반적으로 route 경로는 소문자로 작성합니다. 띄어쓰기는 `-`으로 처리해주시면 됩니다.

ex) https://github.com/wecode-bootcamp-korea

201steve replied:
path 소문자로 수정했습니다!

lang92 commented on src/components/Nav/Nav.js:
```diff
@@ -1,11 +1,53 @@
 import React from 'react';
+import './Nav.scss';
+import { Link } from 'react-router-dom';
```
아래 순서를 참고해서 Import 순서 수정해주세요. 일반적인 convention을 따르는 이유도 있지만 순서만 잘 지켜주셔도 가독성이 좋아집니다.

- React → Library(Package) → Component → 변수 / 이미지 → css 파일(scss 파일)
```suggestion
import { Link } from 'react-router-dom';
import './Nav.scss';
```

리뷰를 남기지 않은 부분들도 찾아서 수정해주세요!

201steve committed:
[Refactor] conflict된index.html 코드  수정

lang92 commented on src/pages/ShopList/ShopListMain.scss:
```diff
@@ -0,0 +1,147 @@
+            &:hover {
+              cursor: pointer;
+            }
+          }
```
위 코드대로 사용하시려고 한다면, 다음과 같이 바꿔볼 수 있을 것 같습니다.
중복되는 코드가 보인다면, 어떻게 간편하게 바꿔볼 수 있을지 생각해보시면 좋을 것 같습니다.
```suggestion
          .fa-star {
            font-size: 30px;
            color: #bcbcbc;
            cursor: pointer;
             
            &.fas {
               color: #ff782a;
             }
           }
```

lang92 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,92 @@
+  }
+
+  componentDidMount() {
+    fetch('http://10.58.0.54:8000/restaurants?category=3')
```
config.js 생성해서 api 주소 따로 관리해주세요!

lang92 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,92 @@
+                //handler
+                likeHandle={this.likeHandler}
+                buttonHandle={this.buttonHandler}
+              />
```
- mock data를 사용하는 목적은,
  - 백엔드 API에서 보내주는 response가 어떤 형태인지 key-value의 형태를 미리 맞춰봄으로써
  - 백엔드 API가 완성되기 전까지, 혹은 이후에도 데이터에 맞게 UI가 의도한대로 구현되는지 확인하고
  - API integration을 수월하게 하기 위함입니다.
- mock data와 실제 데이터의 key-value 형태가 다르다고 해서, 이처럼 형태에 따라 각각 코드를 다르게 짤 필요는 없습니다.
- 동일한 props에 동일한 형태로 전달할 수 있도록, mock data의 key-value 형태를 실제 데이터에 맞게 수정해주세요.

lang92 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,92 @@
+        <Nav />
+        <ShopListHeader />
+        {shopInfo &&
+          shopInfo.map(list => {
```
key prop을 생략하셨네요!
map 메서드로 rendering 시에 key prop이 왜 필요한지 생각해보시고, key prop 작성해주시기 바랍니다.

201steve committed:
[ADD] header에 필터버튼 추가

201steve committed:
[MODIFY] conflict 해결

lang92 commented on src/pages/ShopList/ShopList.js:
```diff
@@ -1,10 +1,92 @@
+    // .then(res => this.setState({ shopInfo: res }));
+  }
+
+  buttonHandler = e => {
+    const changeList = [...this.state.shopInfo];
+    changeList.map(el => {
+      if (Number(e.target.name) === el.id) {
+        console.log('aaa');
+        el.btn_toggle = !el.btn_toggle;
+      }
+    });
+    this.setState({
+      shopInfo: changeList,
+    });
+  };
+
+  likeHandler = e => {
+    const changeList = [...this.state.shopInfo];
+    changeList.map(el => {
+      if (Number(e.target.id) === el.id) {
+        el.is_wished = !el.is_wished;
+      }
+    });
+    this.setState({
+      shopInfo: changeList,
+    });
+  };
```
forEach 메서드는 **현재 배열을 변경**하고, map 메서드는 대신 ​​**새 배열을 반환**한다는 차이점이 있습니다. 그렇기 때문에 지금과 같은 방식의 경우에는 map 메서드가 아닌 forEach 메서드를 사용하셔야 합니다.

map 메서드를 사용하시려면, 
```
  buttonHandler = e => {
    const changeList = this.state.shopInfo.map(el => {
      return Number(e.target.name) === el.id
        ? {
            ...el,
            btn_toggle: !el.btn_toggle,
          }
        : el;
    });
    this.setState({
      shopInfo: changeList,
    });
  };
```
위와 같은 방식으로 작성하시면 됩니다.

lang92 reviewed (commented):
고생하셨습니다 정일님! 리뷰 반영해주세요~!

- self refactoring checklist 7, 8번 참고하셔서 css 속성 순서 일관되게 통일해 주시고, 새로운 선택자 간에는 가독성을 위해 한 줄씩 띄워 주세요!

- 함수명은 함수의 의도와 기능을 명확히할 수 있는 동사형이어야 합니다.
ex) buttonHandler -> handleBtn,
 likeHandler -> handleWishBtn

201steve committed:
[MODIFY] 리뷰 수정사항 반영
